package/upgrade/upgrade_node.sh

```diff
@@ -0,0 +1,322 @@
+  	virtctl stop $repo_vm_name -n harvester-system
+  fi
+}
+
+shutdown_all_vms()
+{
+  kubectl get vmi -A -o json |
+    jq -r '.items[] | [.metadata.name, .metadata.namespace] | @tsv' |
+    while IFS=$'\t' read -r name namespace; do
+      if [ -z "$name" ]; then
+        break
+      fi
+      echo "Stop ${namespace}/${name}"
+      virtctl stop $name -n $namespace
+    done
+}
+
+shutdown_non_migrate_able_vms()
+{
+  # VMs with nodeSelector
+  kubectl get vmi -A -l kubevirt.io/nodeName=$HARVESTER_UPGRADE_NODE_NAME -o json |
+    jq -r '.items[] | select(.spec.nodeSelector != null) | [.metadata.name, .metadata.namespace] | @tsv' |
+    while IFS=$'\t' read -r name namespace; do
+      if [ -z "$name" ]; then
+        break
+      fi
+      echo "Stop ${namespace}/${name}"
+      virtctl stop $name -n $namespace
+    done
+
+  # VMs with nodeAffinity
+  kubectl get vmi -A -l kubevirt.io/nodeName=$HARVESTER_UPGRADE_NODE_NAME -o json |
+    jq -r '.items[] | select(.spec.affinity.nodeAffinity != null) | [.metadata.name, .metadata.namespace] | @tsv' |
+    while IFS=$'\t' read -r name namespace; do
+      if [ -z "$name" ]; then
+        break
+      fi
+      echo "Stop ${namespace}/${name}"
+      virtctl stop $name -n $namespace
+    done
+}
+
+command_prepare()
+{
+  wait_repo
+  detect_repo
+  preload_images
+}
+
+wait_evacuation_pdb_gone()
+{
+  # TODO: fine-tune this to per-VM check
+  until ! kubectl get pdb -o name -A | grep kubevirt-migration-pdb-kubevirt-evacuation-
+  do
+    echo "Waiting for evacuation PDB gone..."
+    sleep 5
+  done
+}
+
+
+command_pre_drain() {
+  shutdown_non_migrate_able_vms
+
+  # Live migrate VMs
+  kubectl taint node $HARVESTER_UPGRADE_NODE_NAME --overwrite kubevirt.io/drain=draining:NoSchedule
+
+  # Wait for VM migrated
+  wait_vms_out
+
+  # KubeVirt's pdb might cause drain fail
+  wait_evacuation_pdb_gone
+}
+
+get_node_rke2_version() {
+  kubectl get node $HARVESTER_UPGRADE_NODE_NAME -o yaml | yq -e e '.status.nodeInfo.kubeletVersion' -
+}
+
+upgrade_rke2() {
+  patch_file=$(mktemp -p $UPGRADE_TMP_DIR)
+
+cat > $patch_file <<EOF
+spec:
+  kubernetesVersion: $REPO_RKE2_VERSION
+  rkeConfig: {}
+EOF
+
+  kubectl patch clusters.provisioning.cattle.io local -n fleet-local --patch-file $patch_file --type merge
+}
+
+wait_rke2_upgrade() {
+  until [ "$(get_node_rke2_version)" = "$REPO_RKE2_VERSION" ]
+  do
+    echo "Waiting for RKE2 to be upgraded..."
+    sleep 5
+  done
+}
+
+rebrand_grub () {
+  mount -o remount,rw $HOST_DIR/run/initramfs/cos-state/
+  chroot $HOST_DIR grub2-editenv /run/initramfs/cos-state/grub_oem_env set "default_menu_entry=$REPO_OS_PRETTY_NAME"
+  mount -o remount,ro $HOST_DIR/host/run/initramfs/cos-state/
+}
+
+upgrade_os() {
+  CURRENT_OS_VERSION=$(source $HOST_DIR/etc/os-release && echo $PRETTY_NAME)
+
+  if [ "$REPO_OS_PRETTY_NAME" = "$CURRENT_OS_VERSION" ]; then
+    echo "Skip upgrading OS. The version version is already \"$CURRENT_OS_VERSION\"."
```

Comment:
Seems like a typo? Not sure if this is the correct word:
```suggestion
    echo "Skip upgrading OS. The Harvester version is already \"$CURRENT_OS_VERSION\"."
```

Reply:
Thanks, fixed.